### PR TITLE
JENKINS-46132 - Download actual proc info from the Linux agents

### PR DESCRIPTION
**Pull request by duemir**
https://issues.jenkins-ci.org/browse/JENKINS-46132

FileContent which was using Master's files for all the agents is
replaced with FilePathContent which loads actual info via remoting.

Started phasing out the Active Instance Helper, since Jenkins Core
dependency is now greater than 1.609

@reviewbybees

### CODE REVIEW

**Comment by duemir** [comment]
This will, unfortunately, disable it for both master and agents. 
I am not ready to lose the info from the master. It is very handy at judging the sizing in performance issue and double checking that arguments were added as suggested.

As a compromise, we can just not capture it from agents at all or split it into another component which will be disabled by default.

**Comment by duemir** [comment]
@stephenc you seem to be an auther of the `FilePathContent` class. Do you mind providing feedback to the Owen's concern?

**Comment by duemir** [comment]
@christ66 can you give your 2c about it?

File: src/main/java/com/cloudbees/jenkins/support/impl/ProcFilesRetriever.java [hunk 1]
```diff
@@ -55,7 +53,7 @@ protected static String getNodeName(Node node) {
 
     @Override
     public void addContents(@NonNull Container container) {
-        Jenkins j = Helper.getActiveInstance();
+        Jenkins j = Jenkins.getActiveInstance();
```
Comment (christ66) [inline src/main/java/com/cloudbees/jenkins/support/impl/ProcFilesRetriever.java]: :NIT: should be part of a different PR (I assume it's safe to use and should just get rid of the helper version and just use the one in the `Jenkins` class.

**Comment by christ66** [comment]
If you would like to do this for `Master` configurations, but make it optional for the `Agent` configuration, my advice is just create 2 different versions of this.

File: src/main/java/com/cloudbees/jenkins/support/impl/ProcFilesRetriever.java (hunk 1 above)
Comment (batmat) [reply src/main/java/com/cloudbees/jenkins/support/impl/ProcFilesRetriever.java]: +1, and possibly more than just NIT because https://github.com/jenkinsci/jenkins/pull/3195... :)

File: src/main/java/com/cloudbees/jenkins/support/impl/ProcFilesRetriever.java [hunk 2]
```diff
@@ -65,11 +63,11 @@ public void addContents(@NonNull Container container) {
 
     protected void addUnixContents(@NonNull Container container, final @NonNull Node node) {
         Computer c = node.toComputer();
-        if (c == null) {
+        if (c == null || c.isOffline()) {
             return;
         }
         // fast path bailout for Windows
-        if (c instanceof SlaveComputer && !Boolean.TRUE.equals(((SlaveComputer) c).isUnix())) {
+        if (!Boolean.TRUE.equals(c.isUnix())) {
```
Comment (batmat) [inline src/main/java/com/cloudbees/jenkins/support/impl/ProcFilesRetriever.java]: This and above seem unrelated?

**Comment by batmat** [comment]
Also, test failure. Retriggering. If this persists, you'll obviously want to wait until it's fixed before merging.

File: src/main/java/com/cloudbees/jenkins/support/impl/ProcFilesRetriever.java (hunk 1 above)
Comment (jglick) [reply src/main/java/com/cloudbees/jenkins/support/impl/ProcFilesRetriever.java]: It may be deprecated in later baselines, but it works as advertised and will continue to.

File: src/main/java/com/cloudbees/jenkins/support/impl/ProcFilesRetriever.java (hunk 2 above)
Comment (duemir) [reply src/main/java/com/cloudbees/jenkins/support/impl/ProcFilesRetriever.java]: You mean any refactoring due to new baseline should be in the separate PR?

**Comment by duemir** [comment]
1. Now there are 4 components instead of 2. Those 2 that are for agents are not selected by default.
2. No more `getInstance` update.

**Comment by duemir** [comment]
3. Rebased on top of the `master`
4. Fixed the test

**Comment by duemir** [comment]
Are you kidding me? I resolved this conflict from your wonderful UI Github, why are you telling me that it's still there!!!

**Comment by duemir** [comment]
All resolved and fixed. Ready for review again.

**Comment by duemir** [comment]
@batmat any luck reviewing this 
@Evildethow @christ66 what do you think about updated version?

File: src/main/java/com/cloudbees/jenkins/support/api/FilePathContent.java [hunk 3]
```diff
@@ -24,11 +24,11 @@
 
 package com.cloudbees.jenkins.support.api;
 
+import com.cloudbees.jenkins.support.SupportLogFormatter;
 import hudson.FilePath;
-import hudson.util.IOException2;
 
-import java.io.IOException;
-import java.io.OutputStream;
+import java.io.*;
```
Comment (batmat) [inline src/main/java/com/cloudbees/jenkins/support/api/FilePathContent.java]: hmmm

File: src/main/java/com/cloudbees/jenkins/support/api/FilePathContent.java [hunk 4]
```diff
@@ -49,7 +53,25 @@ public void writeTo(OutputStream os) throws IOException {
         try {
             file.copyTo(os);
         } catch (InterruptedException e) {
-            throw new IOException2(e);
+            throw new IOException(e);
+        } catch (IOException e) {
+            if (isFileNotFound(e) || isFileNotFound(e.getCause())) {
+                OutputStreamWriter osw = new OutputStreamWriter(os, "utf-8");
+                try {
```
Comment (batmat) [inline src/main/java/com/cloudbees/jenkins/support/api/FilePathContent.java]: `try-with-resources`?

Comment (duemir) [reply src/main/java/com/cloudbees/jenkins/support/api/FilePathContent.java]: the `os` comes from the caller which is supposed to manage its lifecycle. We don't want to close it, even though there are provisions made to prevent `Content` classes from closing this stream.

File: src/main/java/com/cloudbees/jenkins/support/api/FilePathContent.java (hunk 3 above)
Comment (duemir) [reply src/main/java/com/cloudbees/jenkins/support/api/FilePathContent.java]: My IntelliJ settings might be a bit aggressive about wildcard imports. Didn't know if there is some standard regarding this. Is there one?

Comment (christ66) [reply src/main/java/com/cloudbees/jenkins/support/api/FilePathContent.java]: We've been advising to use explicit imports since it makes having class name conflicts easier to resolve.

File: src/main/java/com/cloudbees/jenkins/support/impl/ProcFilesRetriever.java (hunk 2 above)
Comment (christ66) [reply src/main/java/com/cloudbees/jenkins/support/impl/ProcFilesRetriever.java]: Yes. For future PRs.

Comment (christ66) [reply src/main/java/com/cloudbees/jenkins/support/impl/ProcFilesRetriever.java]: It should be refactored when the new baseline is updated.

File: src/main/java/com/cloudbees/jenkins/support/impl/SystemConfiguration.java [hunk 5]
```diff
@@ -46,8 +47,29 @@
...
  */
-@Extension
-public class SystemConfiguration extends ProcFilesRetriever {
+public abstract class SystemConfiguration extends ProcFilesRetriever {
+    @Extension
+    public static class Master extends SystemConfiguration {
+        @Override
+        @NonNull
+        public String getDisplayName() {
+            return "Master system configuration (Linux only)";
```
Comment (christ66) [inline src/main/java/com/cloudbees/jenkins/support/impl/SystemConfiguration.java]: Sine this is only available for linux I think we should add the ability to not show it if it's on windows. I.E.

`@Override public boolean isEnabled() { return (Platform.current() == Platform.UNIX); }`

**Comment by christ66** [comment]
Minor :NITS:. I think it can be merged.

**Comment by duemir** [comment]
@christ66 I addressed the imports comment. Would appreciate if you can merge it.

**Comment by duemir** [comment]
5th most commented PR. Just a handful more comments and I will win 😆

**Comment by duemir** [comment]
I don't think I have ever had a PR in review for almost half a year. Setting personal records on all fronts 🏆
